### PR TITLE
Record column lineage details 

### PR DESCRIPTION
**Pull request by Praveen2112**
Overrides https://github.com/trinodb/trino/pull/7354

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 1]
```diff
@@ -957,6 +960,38 @@ public void addColumnMask(Table table, String column, Expression mask)
...
     }
 
+    public void addOriginColumnDetail(Field field, Collection<ColumnEntry> columnEntries)
+    {
+        ImmutableSet.Builder<ColumnEntry> columnEntryBuilder = ImmutableSet.builder();
+        columnEntryBuilder.addAll(columnEntries);
+        if (field.getOriginTable().isPresent()) {
+            columnEntryBuilder.add(new ColumnEntry(field.getOriginTable().get(), field.getOriginColumnName().get()));
+        }
+        originColumnDetails.putAll(field, columnEntries);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:970]: Use `columnEntryBuilder.builder()`, It look like you have a test missing.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 2]
```diff
@@ -957,6 +960,38 @@ public void addColumnMask(Table table, String column, Expression mask)
                 .collect(toImmutableList());
     }
 
+    public void addOriginColumnDetail(Field field, Collection<ColumnEntry> columnEntries)
+    {
+        ImmutableSet.Builder<ColumnEntry> columnEntryBuilder = ImmutableSet.builder();
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:965]: names `columnEntryBuilder` and `columnEntries` are not good, it is not easy to see the difference and decide which one should be used.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/OutputColumn.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+@Immutable
+public final class OutputColumn
+{
+    private final Column column;
+    private final Set<ColumnEntry> referencedColumns;
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/OutputColumn.java:32]: This sounds like a bad name. It is like a source column, wdyt?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 4]
```diff
@@ -29,7 +29,9 @@
 import io.trino.spi.connector.ConnectorFactory;
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:29]: add a test where column has multiple referenced columns (source columns)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 5]
```diff
@@ -480,8 +482,10 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
                     "INSERT",
                     targetTable,
                     Optional.empty(),
-                    Optional.of(insertColumns.stream()
-                            .map(insertColumn -> new Column(insertColumn, tableMetadata.getColumn(insertColumn).getType().toString()))
+                    Optional.of(Streams.zip(
+                            insertColumns.stream(),
+                            queryScope.getRelationType().getVisibleFields().stream(),
+                            (column, field) -> new OutputColumn(new Column(column, field.getType().toString()), ImmutableSet.copyOf(analysis.getOriginColumns(field))))
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:488]: Extract this as method, it is used frequently below

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 6]
```diff
@@ -1973,6 +1981,7 @@ public static ExpressionAnalysis analyzeExpression(
         analyzer.analyze(expression, scope);
 
         updateAnalysis(analysis, analyzer, session, accessControl);
+        analysis.addFieldLineage(expression, analyzer.getSourceFields());
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:1984]: There are other method which should be updated, see `analyzeWindow` or `analyzeExpressions` where this is not added.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 7]
```diff
@@ -29,7 +29,9 @@
...
 import io.trino.spi.connector.ConnectorViewDefinition;
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:30]: Do you have test for `CREATE TABLE LIKE` where we should see that columns have nice source columns set.

Comment (Praveen2112) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:30]: In case of `LIKE` we don't capture the data right ? We just reuse the name and type.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/OutputColumn.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+import io.trino.sql.analyzer.Analysis.SourceColumn;
+
+import javax.annotation.concurrent.Immutable;
+
+import java.util.Objects;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+@Immutable
```
Comment (skrzypo987) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/OutputColumn.java:29]: Does this annotation do anything or is it just information?

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 9]
```diff
@@ -1013,21 +1024,22 @@ protected Scope visitCreateMaterializedView(CreateMaterializedView node, Optiona
 
             Scope queryScope = analyzer.analyze(node.getQuery(), scope);
 
+            validateColumns(node, queryScope.getRelationType());
```
Comment (skrzypo987) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:1027]: Why is this moved?

**Review by skrzypo987 — COMMENTED** [review]
Skimmed, looks ok

File: core/trino-main/src/main/java/io/trino/sql/analyzer/OutputColumn.java (hunk 8 above)
Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/OutputColumn.java:29]: It has some compile time checks I guess.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 4 above)
Comment (Praveen2112) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:29]: I guess it is already added.

**Comment by Praveen2112** [comment]
@kokosing , @skrzypo987 AC

File: core/trino-main/src/main/java/io/trino/sql/analyzer/OutputColumn.java (hunk 8 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/OutputColumn.java:29]: I was under the impression it is rarely used but I was wrong. It is used throughout the project.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 10]
```diff
@@ -957,6 +959,33 @@ public void addColumnMask(Table table, String column, Expression mask)
...
+    {
+        originColumnDetails.putAll(field, sourceColumn);
+    }
+
+    public Collection<SourceColumn> getSourceColumns(Field field)
+    {
+        return originColumnDetails.get(field);
+    }
+
+    public void addFieldLineage(Expression expression, Collection<Field> fields)
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:972]: addExpressionFields. I think lineage is just a one of the use cases

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 11]
```diff
@@ -957,6 +959,33 @@ public void addColumnMask(Table table, String column, Expression mask)
...
+    {
+        fieldLineage.putAll(NodeRef.of(expression), fields);
+    }
+
+    public Collection<SourceColumn> getSourceColumnsForExpression(Expression expression)
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:977]: getExpressionSourceColumns

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 12]
```diff
@@ -957,6 +959,33 @@ public void addColumnMask(Table table, String column, Expression mask)
...
+    {
+        return fieldLineage.get(NodeRef.of(expression)).stream()
+                .flatMap(field -> getSourceColumns(field).stream())
+                .collect(toImmutableSet());
+    }
+
+    public Collection<SourceColumn> getOriginColumnDetails(Field field)
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:984]: It is redundant. There is `getSourceColumns`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 13]
```diff
@@ -1479,6 +1508,48 @@ public Scope getAccessControlScope()
         }
     }
 
+    public static class SourceColumn
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:1511]: You can add a convenience method to convert this class to `io.trino.spi.eventlistener.ColumnDetail`. So you could use it `io.trino.event.QueryMonitor#getQueryIOMetadata`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java [hunk 14]
```diff
@@ -957,6 +959,33 @@ public void addColumnMask(Table table, String column, Expression mask)
...
     }
 
+    public void addSourceColumns(Field field, Collection<SourceColumn> sourceColumn)
+    {
+        originColumnDetails.putAll(field, sourceColumn);
+    }
+
+    public Collection<SourceColumn> getSourceColumns(Field field)
+    {
+        return originColumnDetails.get(field);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/Analysis.java:969]: return `ImmutableSet.copyOf`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 15]
```diff
@@ -743,7 +750,9 @@ protected Scope visitCreateTableAsSelect(CreateTableAsSelect node, Optional<Scop
                     if (field.getType().equals(UNKNOWN)) {
                         throw semanticException(COLUMN_TYPE_UNKNOWN, node, "Column type is unknown at position %s", queryScope.getRelationType().indexOf(field) + 1);
                     }
-                    columns.add(new ColumnMetadata(node.getColumnAliases().get().get(aliasPosition).getValue(), field.getType()));
+                    String columnName = node.getColumnAliases().get().get(aliasPosition).getValue();
+                    columns.add(new ColumnMetadata(columnName, field.getType()));
+                    outputColumns.add(new OutputColumn(new Column(columnName, field.getType().toString()), ImmutableSet.copyOf(analysis.getSourceColumns(field))));
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:755]: make `getSourceColumns` to do a `copyOf`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 16]
```diff
@@ -752,6 +761,10 @@ protected Scope visitCreateTableAsSelect(CreateTableAsSelect node, Optional<Scop
                 columns.addAll(queryScope.getRelationType().getVisibleFields().stream()
                         .map(field -> new ColumnMetadata(field.getName().get(), field.getType()))
                         .collect(toImmutableList()));
+                queryScope.getRelationType().getVisibleFields()
+                        .forEach(field -> outputColumns.add(new OutputColumn(
+                                new Column(field.getName().get(),
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:766]: nit: put each arg into separate field

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/OutputColumnMetadata.java [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+import java.util.Objects;
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+public class OutputColumnMetadata
+{
+    private final String columnName;
+    private final Set<ColumnDetail> referencedColumns;
```
Comment (kokosing) [inline core/trino-spi/src/main/java/io/trino/spi/eventlistener/OutputColumnMetadata.java:26]: sourceColumn?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 18]
```diff
@@ -623,27 +641,95 @@ public void testOutputColumnsForCreateTableAsSelect()
...
+                .containsExactly(
+                        new OutputColumnMetadata("custkey", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "orders", "custkey"))),
+                        new OutputColumnMetadata("col2", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "nation", "nationkey"))));
+    }
+
+    @Test
+    public void testOutputColumnsForCreateTableAsSelectWithCorrelatedQueries()
+            throws Exception
+    {
+        runQueryAndWaitForEvents("CREATE TABLE mock.default.create_new_table_with_un_correlated_queries AS SELECT name, (SELECT sum(acctbal) FROM supplier WHERE supplier.nationkey=nation.nationkey) as col2 FROM nation",
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:716]: Please extract some utility assertion method
```
assertColumnLinage(
    "SELECT name, (SELECT sum(acctbal) FROM supplier WHERE supplier.nationkey=nation.nationkey) as col2 FROM nation",
    columnLinage("name", "tpch.tiny.nation.name"),
    columnLinage("col2", "tpch.tiny.supplier.acctbal"));
```

Please run the above "SELECT.." statement for CTAS, INSERT, UPDATE, MERGE. The last two might require more tricks to make it possible. For MERGE you should expect that statement is going to fail, but once we add support one would need to update this.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 19]
```diff
@@ -623,27 +641,95 @@ public void testOutputColumnsForCreateTableAsSelect()
...
+                2);
+        QueryCompletedEvent event = getOnlyElement(generatedEvents.getQueryCompletedEvents());
+        assertThat(event.getIoMetadata().getOutput().get().getColumns().get())
+                .containsExactly(
+                        new OutputColumnMetadata("custkey", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "orders", "custkey"))),
+                        new OutputColumnMetadata("col2", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "nation", "nationkey"))));
+    }
+
+    @Test
+    public void testOutputColumnsForCreateTableAsSelectWithCorrelatedQueries()
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:713]: I don't see tests for HAVING, LIMIT, ORDER BY and lambdas.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 20]
```diff
@@ -554,8 +558,10 @@ protected Scope visitRefreshMaterializedView(RefreshMaterializedView refreshMate
                     "REFRESH MATERIALIZED VIEW",
                     targetTable,
                     Optional.empty(),
-                    Optional.of(insertColumns.stream()
-                            .map(insertColumn -> new Column(insertColumn, tableMetadata.getColumn(insertColumn).getType().toString()))
+                    Optional.of(Streams.zip(
+                            insertColumns.stream(),
+                            queryScope.getRelationType().getVisibleFields().stream(),
+                            this::createOutputColumn)
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:564]: The new method uses `field.getType()` for the `Column` type, instead of `tableMetadata.getColumn(insertColumn).getType()`. The types of are not necessarily the same (see `typesMatchForInsert `)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 21]
```diff
@@ -480,8 +482,10 @@ protected Scope visitInsert(Insert insert, Optional<Scope> scope)
                     "INSERT",
                     targetTable,
                     Optional.empty(),
-                    Optional.of(insertColumns.stream()
-                            .map(insertColumn -> new Column(insertColumn, tableMetadata.getColumn(insertColumn).getType().toString()))
+                    Optional.of(Streams.zip(
+                            insertColumns.stream(),
+                            queryScope.getRelationType().getVisibleFields().stream(),
+                            this::createOutputColumn)
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:488]: The new method uses field.getType() for the Column type, instead of tableMetadata.getColumn(insertColumn).getType(). The types of are necessarily the same (see typesMatchForInsert)

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 20 above)
Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:564]: Yeah !! But we need also `Field` for fetching the lineage. So instead of fetching from the `tableMetadata` we thought of fetching it from `Field`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 21 above)
Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:488]: https://github.com/trinodb/trino/pull/7465#discussion_r606285522

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 20 above)
Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:564]: All right, I understand that you need to fetch the source columns from field. But besides that, there's another unrelated detail that changed. It is the Type used for the `Column` construction. It should be taken from the column metadata, not from the field.

**Comment by Praveen2112** [comment]
@kasiafi Thanks for the review. AC

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 20 above)
Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java]: Oh okay. Have addressed it. Thanks for the pointer.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 22]
```diff
@@ -3671,6 +3702,11 @@ private Scope createScope(Optional<Scope> parentScope)
 
             return scopeBuilder;
         }
+
+        private OutputColumn createOutputColumn(Field field)
+        {
+            return new OutputColumn(new Column(field.getName().orElseThrow(), field.getType().toString()), analysis.getSourceColumns(field));
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3708]: Is it checked on some earlier stage of the analysis that all fields have names? If they don't, that should be a `SematicException` probably.

Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3708]: We make sure that we create the ouputColumn after all the validation

Comment (kasiafi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3708]: Indeed. So maybe just do an unchecked `get()`.

File: core/trino-main/src/test/java/io/trino/sql/analyzer/TestOutput.java [hunk 23]
```diff
@@ -29,7 +30,7 @@
     @Test
     public void testRoundTrip()
     {
-        Output expected = new Output("connectorId", "schema", "table", Optional.of(ImmutableList.of(new Column("column", "type"))));
+        Output expected = new Output("connectorId", "schema", "table", Optional.of(ImmutableList.of(new OutputColumn(new Column("column", "type"), ImmutableSet.of()))));
```
Comment (kasiafi) [inline core/trino-main/src/test/java/io/trino/sql/analyzer/TestOutput.java:33]: Test it with some source column.

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/OutputColumnMetadata.java [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+        if ((obj == null) || (getClass() != obj.getClass())) {
+            return false;
+        }
+        OutputColumnMetadata outputColumnMetadata = (OutputColumnMetadata) obj;
```
Comment (kasiafi) [inline core/trino-spi/src/main/java/io/trino/spi/eventlistener/OutputColumnMetadata.java:60]: `outputColumnMetadata` -> `other`

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java (hunk 22 above)
Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3708]: From `Optional#get` java docs 
> The preferred alternative to this method is orElseThrow().

File: core/trino-spi/src/main/java/io/trino/spi/eventlistener/OutputColumnMetadata.java (hunk 24 above)
Comment (Praveen2112) [reply core/trino-spi/src/main/java/io/trino/spi/eventlistener/OutputColumnMetadata.java]: AC

File: core/trino-main/src/test/java/io/trino/sql/analyzer/TestOutput.java (hunk 23 above)
Comment (Praveen2112) [reply core/trino-main/src/test/java/io/trino/sql/analyzer/TestOutput.java]: AC

File: core/trino-main/src/main/java/io/trino/event/QueryMonitor.java [hunk 25]
```diff
@@ -363,15 +366,21 @@ private static QueryIOMetadata getQueryIOMetadata(QueryInfo queryInfo)
                     .map(TableFinishInfo.class::cast)
                     .findFirst();
 
+            Optional<List<OutputColumnMetadata>> outputColumnsMetadata = queryInfo.getOutput().get().getColumns()
+                    .map(outputColumn -> outputColumn.stream()
```
Comment (kasiafi) [inline core/trino-main/src/main/java/io/trino/event/QueryMonitor.java:370]: `outputColumns` or just `columns`

Comment (Praveen2112) [reply core/trino-main/src/main/java/io/trino/event/QueryMonitor.java]: AC

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 26]
```diff
@@ -468,7 +486,7 @@ public void testReferencedTablesWithColumnMask()
 
         column = table.getColumns().get(1);
         assertThat(column.getColumn()).isEqualTo("test_bigint");
-        assertThat(column.getMasks()).isEmpty();
+        assertThat(column.getMasks()).hasSize(0);
```
Comment (kasiafi) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:489]: Why did it change? Is it just a refactor? If so, could you please extract it to another commit? Also, there are multiple occurrences of `assertThat(...).isEmpty();` in this class.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 27]
```diff
@@ -623,27 +641,95 @@ public void testOutputColumnsForCreateTableAsSelect()
...
+                        new OutputColumnMetadata("name", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "nation", "name"))),
+                        new OutputColumnMetadata("nationkey", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "nation", "nationkey"))));
     }
 
     @Test
     public void testOutputColumnsForCreateTableAsSelectWithAliasedColumn()
             throws Exception
     {
-        runQueryAndWaitForEvents("CREATE TABLE mock.default.create_new_table(aliased_bigint, aliased_varchar) AS SELECT name, nationkey as keynation FROM nation", 2);
+        runQueryAndWaitForEvents("CREATE TABLE mock.default.create_new_table(aliased_varchar, aliased_bigint) AS SELECT concat(name, comment), nationkey as keynation FROM nation", 2);
```
Comment (kasiafi) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:653]: Why did the order change?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 28]
```diff
@@ -623,27 +641,95 @@ public void testOutputColumnsForCreateTableAsSelect()
...
+            throws Exception
+    {
+        runQueryAndWaitForEvents("CREATE TABLE mock.default.create_new_table_with_aggregation(aliased_aggregation) AS SELECT count(*) FROM orders GROUP BY custkey", 2);
+        QueryCompletedEvent event = getOnlyElement(generatedEvents.getQueryCompletedEvents());
+        assertThat(event.getIoMetadata().getOutput().get().getColumns().get())
+                .containsExactly(new OutputColumnMetadata("aliased_aggregation", ImmutableSet.of()));
+    }
+
+    @Test
+    public void testOutputColumnsForCreateTableAsSelectWithWindowFunction()
```
Comment (kasiafi) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:682]: Please add a similar test with `WINDOW` clause instead of in-line window specification.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 29]
```diff
@@ -698,6 +787,9 @@ public void testCreateTableLike()
         assertThat(event.getIoMetadata().getOutput().get().getSchema()).isEqualTo("default");
         assertThat(event.getIoMetadata().getOutput().get().getTable()).isEqualTo("create_simple_table");
         assertThat(event.getIoMetadata().getOutput().get().getColumns().get())
-                .containsExactly("test_column", "test_varchar", "test_bigint");
+                .containsExactly(
+                        new OutputColumnMetadata("test_column", ImmutableSet.of()),
+                        new OutputColumnMetadata("test_varchar", ImmutableSet.of()),
+                        new OutputColumnMetadata("test_bigint", ImmutableSet.of()));
```
Comment (kasiafi) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:793]: Please add tests with column references in:
- `WITH` query
- aggregation's filter, aggregation's ORDER BY
- ORDER BY
- WHERE

A question: why are the grouping columns of an aggregation not recorded?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 26 above)
Comment (Praveen2112) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java]: It messed up during rebase

**Comment by Praveen2112** [comment]
@kokosing , @kasiafi AC

**Review by kokosing — APPROVED** [review]

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 30]
```diff
@@ -617,33 +635,159 @@ public void testOutputStats()
...
+            throws Exception
+    {
+        assertColumnLineage(
+                "SELECT name as test_varchar, nationkey as test_bigint FROM nation ORDER BY comment",
+                new OutputColumnMetadata("test_varchar", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "nation", "name"))),
+                new OutputColumnMetadata("test_bigint", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "nation", "nationkey"))));
+    }
+
+    @Test
+    public void testOutputColumnsWithAggregation()
```
Comment (kokosing) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:714]: Please also test: https://trino.io/docs/current/functions/aggregate.html#filtering-during-aggregation

`CASE` and `IF` expressions.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java [hunk 31]
```diff
@@ -976,7 +986,7 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
                 ResolvedWindow window = getResolvedWindow.apply(node);
                 checkState(window != null, "no resolved window for: " + node);
 
-                analyzeWindow(window, context, (Node) node.getWindow().get());
+                analyzeWindow(window, new StackableAstVisitorContext<>(context.getContext().skipTrackingSourceColumns()), (Node) node.getWindow().get());
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java:989]: This is not only about `WINDOW`. It also would need to be handled in https://trino.io/docs/current/functions/aggregate.html#filtering-during-aggregation (where filter is skipped) and `CASE` or `IF` (where condition is skipped). Also there might be other like this in future.

After a second though, I think we should just put all `WINDOW` arguments as source columns for output expression. This makes reasoning easier, code simpler. It will also make it more straightforward for `CASE` and `IF`.

**Comment by Praveen2112** [comment]
@kokosing Added tests

**Review by kokosing — APPROVED** [review]

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 32]
```diff
@@ -674,33 +692,244 @@ public void testOutputStats()
...
+    {
+        assertColumnLineage(
+                "SELECT orderstatus as test_varchar, sum(totalprice) OVER (w ORDER BY custkey ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW) AS test_bigint FROM orders WINDOW w AS (PARTITION BY custkey)",
+                new OutputColumnMetadata("test_varchar", ImmutableSet.of(new ColumnDetail("tpch", "tiny", "orders", "orderstatus"))),
+                new OutputColumnMetadata(
+                        "test_bigint",
+                        ImmutableSet.of(
+                                new ColumnDetail("tpch", "tiny", "orders", "totalprice"),
+                                new ColumnDetail("tpch", "tiny", "orders", "custkey"))));
+    }
```
Comment (kasiafi) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:893]: Use a different column for window partitioning.

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 33]
```diff
@@ -674,33 +692,244 @@ public void testOutputStats()
...
+
+    @Test
+    public void testOutputColumnsWithWindowClause()
```
Comment (kasiafi) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:896]: Please add a test where some output field does not depend on any source columns (e.g. a constant expression).

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java [hunk 34]
```diff
@@ -674,33 +692,244 @@ public void testOutputStats()
...
+                new OutputColumnMetadata(
+                        "test_bigint",
+                        ImmutableSet.of(
+                                new ColumnDetail("tpch", "tiny", "orders", "totalprice"),
+                                new ColumnDetail("tpch", "tiny", "orders", "custkey"),
+                                new ColumnDetail("tpch", "tiny", "orders", "orderdate"))));
+    }
+
+    @Test
+    public void testOutputColumnsWithPartialWindowClause()
```
Comment (kasiafi) [inline testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:882]: I'm not sure we want to differentiate between in-line window specification and WINDOW clause. Especially that those can be mixed within a single function call.
Currently, in-line window specification reports source columns, and WINDOW clause doesn't.

**Review by kasiafi — COMMENTED** [review]
A couple of comments.

I went through the code. However, I could not make sure if all sites are covered which should expose source columns. The same about the test coverage.

**Comment by Praveen2112** [comment]
> I could not make sure if all sites are covered which should expose source columns. The same about the test coverage.

I guess we could cover all the sites in the incremental way. It is kind of experimental for now. WDYT ?

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 34 above)
Comment (Praveen2112) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:882]: This kind of a limitation for now. The long term plan is to report those columns too.

**Comment by Praveen2112** [comment]
@kasiafi AC

File: testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java (hunk 33 above)
Comment (Praveen2112) [reply testing/trino-tests/src/test/java/io/trino/execution/TestEventListenerBasic.java:906]: Added a separate test for it - testOutputColumnsForSelectWithConstantExpression

**Comment by kasiafi** [comment]
Looks good to me, provided that a more consistent column reporting will be added as a follow-up.

**Comment by Praveen2112** [comment]
> provided that a more consistent column reporting will be added as a follow-up.
Sure thing

**Comment by rsaw4** [comment]
@Praveen2112 how to view the lineage? what is sample output for a nested create view, ie viewA which selects from viewB which selects from viewC

**Comment by wangqinghuan** [comment]
Same question with @rsaw4. How to view the lineage details? @Praveen2112

**Comment by amalakar** [comment]
@Praveen2112 we at are trying to leverage this lineage and found that when some some expression is being used source columns are not being captured. For a query like:

```sql
insert into temp.new_intents

SELECT
  intents.start_at as occurred_at,
  at_timezone(intents.start_at at time zone 'UTC', regions.timezone) as occurred_at_local,
  substr(intents.last_pricing_origin_geohash_12, 1, 6) as intent_geohash
FROM core.intents intents
  JOIN core.regions regions 
     ON regions.region = intents.pricing_region

```

Lineage I am seeing is after light transformation of the lineage we get via the QueryIOMetadata:

```json
{
  "hive.temp.new_intents.occurred_at": [
    {
      "columnName": "hive.core.intents.start_at"
    }
  ],
  "hive.temp.new_intents.occurred_at_local": [],
  "hive.temp.new_intents.intent_geohash": []
}
```

The transformation code looks like:

```java
  public Optional<String> getColumnLineage(QueryIOMetadata ioMetadata) {
    Map<String, List<UpstreamColumn>> lineage = new HashMap<>();

    if (ioMetadata.getOutput().isPresent()) {
      QueryOutputMetadata outputMetadata = ioMetadata.getOutput().get();
      if (outputMetadata.getColumns().isPresent()) {
        List<OutputColumnMetadata> outputColumns = outputMetadata.getColumns().get();
        for (OutputColumnMetadata outputColumn : outputColumns) {
          List<UpstreamColumn> upstreamColumns =
              outputColumn.getSourceColumns().stream()
                  .map(col -> new UpstreamColumn(getQualifiedColumnName(col)))
                  .collect(Collectors.toList());
          String outputColumnName =
              String.format(
                  "%s.%s", getQualifiedTableName(outputMetadata), outputColumn.getColumnName());
          lineage.put(outputColumnName, upstreamColumns);
        }
      } 
...
  }
```

**Comment by amalakar** [comment]
Created: https://github.com/trinodb/trino/issues/10272
